### ISSUE TITLE
New trait for types which can report memory usage

### ISSUE THREAD

**Issue by windelbouwman** [rust-lang/rust#63073]
Hi all,

In the course of implementing the sizeof operator in rustpython, we stumbled upon a feature we possibly lack: the ability to request the true memory usage of a struct. Meaning, the memory usage of the struct itself, as reported by `std::mem::size_of` and friends, but also the size of the dynamically allocated heap structures, such as `Vec`.

Our current approach is to dive into the inner workings of the bigint type, but this is probably undesirable, since the resulting code is unsafe.

The proposed change can be viewed here:

https://github.com/RustPython/RustPython/pull/1172/files

My request, would be to have a trait in the `alloc` crate or `std::mem` crate, which can be implemented by other packages, which support the reporting of dynamic memory by an object.

Example of the trait:

```rust
trait DynamicSizeOf {
    /// Returns the memory used by this type, including all its nested fields
    /// which are dynamically allocated.
    fn sizeof(&self) -> usize
}

impl DynamicSizeOf for BigInt {
    fn sizeof(&self) -> usize {
         // arbitrarily deep query on sub structures memory usage.
    }
}

```

Please also see the related issue here: https://github.com/rust-num/num-bigint/issues/98

Is this a good idea? Are there other options available to achieve the same behavior?

@askaholic and @cuviper what are your opinions on this topic? Do you see other ideas?

**Comment by ishitatsuyuki** [comment]
This idea doesn’t seem new: there are several implementation but not for serious use.

- [syn’s procedural macro example](https://github.com/dtolnay/syn/blob/master/examples/heapsize/README.md)
- [Servo’s crate (and a link to the implementation that is actually used in Servo)](https://github.com/servo/heapsize)

**Comment by windelbouwman** [comment]
Hi @ishitatsuyuki , 

Thank you for your reply! It's good to know that the idea is not new.

So the question might be: does this type of trait belong in the standard library under the `std::mem` or `std::alloc` namespace?

**Comment by ishitatsuyuki** [comment]
My opinion is that this functionality does not belong to std. Rust std only include additional functionality when there are well-known needs and a consensus on API (basically what the RFC process demands).

**Comment by Mark-Simulacrum** [comment]
Yes, this does not seem like a good fit for std, as in particular there is not too much advantage in placing it there (and certainly I imagine there are some open questions in the design space, e.g., should we return the allocated size for a container, the usable size, or the used size?)
